### ISSUE TITLE
slow HDFS commands; possible improvements using snakebite

### ISSUE THREAD

**Issue by matteodellamico** [Yelp/mrjob#998]
Hello,

in our workload where several short jobs are run in a sequence, MrJob spends a lot of time dealing with the HDFS command line, copying data over it. This has the overhead of some seconds each time a JVM is spawned to copy the needed python archives to HDFS.

Could integrating with the snakebite library (https://github.com/spotify/snakebite) be a good solution to speed up the system?

**Comment by tarnfeld** [comment]
Last time I checked snakebite didn't support any write operations, which are required for a lot of HDFS operations in mrjob. If you're using Hadoop 2, we wrote a filesystem implementation based around the WebHDFS API for the PyFS library that's available in https://github.com/duedil-ltd/pyfilesystem/pull/2, we have a wrapper implementation for PyFS which makes an mrjob-fs compatible API. We've been using this setup in production for quite some time now, got some pretty crazy speedups from it.

We can probably clean up the latter and put in a patch to mrjob if this makes sense for the project, but it does basically mean the HDFS filesystem uses a completely different underlying filesystem abstraction, which might be a bit too much bloat.

**Comment by davidmarin** [comment]
@tarnfeld, thanks for clueing me in about PyFS (and WebHDFS).

mrjob is mostly copying files to the same directory; I bet we could get the same speedups just by `put`ing multiple files in one command.

**Comment by tarnfeld** [comment]
Yeah it's without a doubt the JVM overhead, no question about it.. i'm sure lumping together calls to `put` would give a useful speed up.

**Comment by davidmarin** [comment]
@tarnfeld also I'm realizing I should take a closer look at https://github.com/duedil-ltd/mrjob; I'm guessing there's a lot of interesting stuff in your fork. Any suggestions about where to start?

And v0.4.3 is out now, if you want to merge it in!

**Comment by tarnfeld** [comment]
> And v0.4.3 is out now, if you want to merge it in!

Yeah we're running a bit behind so will need to comb through the new changes and merge them in.

> I'm guessing there's a lot of interesting stuff in your fork. Any suggestions about where to start?

Most of the interesting stuff is not in the public repo (though we don't have an internal fork of mrjob) but there are a few interesting things, [this is the diff](https://github.com/Yelp/mrjob/compare/master...duedil-ltd:master). Our changes are mostly around things like local job running, support for Hadoop 2 and various other S3+Hadoop implementations. We've got a few changes to make job names friendlier and also some patches to the mrjob filesystem classes.
